### PR TITLE
consistent hash load balance

### PR DESCRIPTION
**Pull request by imxyb**
Signed-off-by: imxyb <xyb4638@gmail.com>

**What this PR does**:
add consistent hash load balance
**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: cluster/loadbalance/consistent_hash.go [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+	"crypto/md5"
+	"encoding/json"
+	"fmt"
+	"hash/crc32"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/apache/dubbo-go/cluster"
```
Comment (hxmhlt) [inline cluster/loadbalance/consistent_hash.go]: Pls split import

File: cluster/loadbalance/consistent_hash.go [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+	"github.com/apache/dubbo-go/protocol"
+)
+
+const (
+	ConsistentHash = "consistenthash"
+	HashNodes      = "hash.nodes"
+	HashArguments  = "hash.arguments"
+)
+
+var selectors = make(map[string]*ConsistentHashSelector)
```
Comment (hxmhlt) [inline cluster/loadbalance/consistent_hash.go]: var ()

File: cluster/loadbalance/consistent_hash.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+	virtualInvokers map[uint32]protocol.Invoker
+	keys            Uint32Slice
+	argumentIndex   []int
+}
+
+func NewConsistentHashSelector(invokers []protocol.Invoker, methodName string,
+	hashCode uint32) *ConsistentHashSelector {
+	if len(invokers) == 0 {
+		panic("none of invokers")
+	}
```
Comment (hxmhlt) [inline cluster/loadbalance/consistent_hash.go]: Don't  panic here. len(invokers) has been validated by func (invoker *baseClusterInvoker) checkInvokers in base_cluster_invoker.go

File: cluster/loadbalance/consistent_hash.go (hunk 1 above)
Comment (imxyb) [reply cluster/loadbalance/consistent_hash.go]: done

File: cluster/loadbalance/consistent_hash.go (hunk 3 above)
Comment (imxyb) [reply cluster/loadbalance/consistent_hash.go]: done

File: cluster/loadbalance/consistent_hash.go (hunk 2 above)
Comment (imxyb) [reply cluster/loadbalance/consistent_hash.go]: done

File: cluster/loadbalance/consistent_hash.go [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+)
+
+func init() {
+	extension.SetLoadbalance(ConsistentHash, NewConsistentHash)
+}
+
+type ConsistentHashLoadBalance struct {
+}
+
+func NewConsistentHash() cluster.LoadBalance {
```
Comment (zouyx) [inline cluster/loadbalance/consistent_hash.go]: why not private method？

File: cluster/loadbalance/consistent_hash_test.go [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+ * limitations under the License.
+ */
+
+package loadbalance
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/suite"
```
Comment (fangyincheng) [inline cluster/loadbalance/consistent_hash_test.go]: format please

Comment (imxyb) [reply cluster/loadbalance/consistent_hash_test.go]: done

**Comment by imxyb** [comment]
@fangyincheng @zouyx @hxmhlt PTAL

File: cluster/loadbalance/consistent_hash.go [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+	methodName := invocation.MethodName()
+	key := invokers[0].GetUrl().ServiceKey() + "." + methodName
+
+	// hash the invokers
+	bs := make([]byte, 0)
+	for _, invoker := range invokers {
+		b, err := json.Marshal(invoker)
+		if err != nil {
+			panic(fmt.Sprintf("parse json failed:%s", err.Error()))
+		}
```
Comment (hxmhlt) [inline cluster/loadbalance/consistent_hash.go]: do not panic

File: cluster/loadbalance/consistent_hash.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+	selector.virtualInvokers = make(map[uint32]protocol.Invoker)
+	selector.hashCode = hashCode
+	url := invokers[0].GetUrl()
+	selector.replicaNum = int(url.GetMethodParamInt(methodName, HashNodes, 160))
+	indices := re.Split(url.GetMethodParam(methodName, HashArguments, "0"), -1)
+	for _, index := range indices {
+		i, err := strconv.Atoi(index)
+		if err != nil {
+			panic(err)
+		}
```
Comment (hxmhlt) [inline cluster/loadbalance/consistent_hash.go]: do not panic

File: cluster/loadbalance/consistent_hash.go (hunk 6 above)
Comment (imxyb) [reply cluster/loadbalance/consistent_hash.go]: @hxmhlt change to 'continue' or 'log.Warnf'?

Comment (fangyincheng) [reply cluster/loadbalance/consistent_hash.go]: @imxyb we can't break the program when a error occur, so I think you need to handle the possible error.

Comment (imxyb) [reply cluster/loadbalance/consistent_hash.go]: @fangyincheng i know, but the `Select` method not return the error...

Comment (fangyincheng) [reply cluster/loadbalance/consistent_hash.go]: return nil？

File: go.mod [hunk 8]
```diff
@@ -51,3 +51,5 @@ require (
 	google.golang.org/grpc v1.22.1
 	gopkg.in/yaml.v2 v2.2.2
 )
+
+go 1.13
```
Comment (zouyx) [inline go.mod]: I think you should delete this block , No description about this block from golang.org ,I don’t know how will it use in the future

Comment (imxyb) [reply go.mod]: ok, i will remove it. This block added by my goland...😅

Comment (imxyb) [reply go.mod]: done

File: cluster/loadbalance/consistent_hash.go (hunk 7 above)
Comment (imxyb) [reply cluster/loadbalance/consistent_hash.go]: done

File: cluster/loadbalance/consistent_hash.go (hunk 6 above)
Comment (imxyb) [reply cluster/loadbalance/consistent_hash.go]: done

File: cluster/loadbalance/consistent_hash.go [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+type ConsistentHashSelector struct {
+	hashCode        uint32
+	replicaNum      int
+	virtualInvokers map[uint32]protocol.Invoker
+	keys            Uint32Slice
+	argumentIndex   []int
+}
+
+func newConsistentHashSelector(invokers []protocol.Invoker, methodName string,
+	hashCode uint32) *ConsistentHashSelector {
```
Comment (AlexStocks) [inline cluster/loadbalance/consistent_hash.go]: add a blank line pls.

Comment (imxyb) [reply cluster/loadbalance/consistent_hash.go]: done

**Comment by AlexStocks** [comment]
LGTM

**Pull request merged**